core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -395,31 +382,100 @@ private void enterIdleMode() {
     channelStateManager.gotoState(IDLE);
   }
 
-  // Must be run from channelExecutor
-  private void cancelIdleTimer() {
-    if (idleModeTimerFuture != null) {
-      idleModeTimerFuture.cancel(false);
-      idleModeTimer.cancelled = true;
-      idleModeTimerFuture = null;
-      idleModeTimer = null;
+  static final class Rescheduler {
+    private final Runnable runnable;
+    private final ChannelExecutor exec;
+    private final ScheduledExecutorService scheduler;
+
+    private ScheduledFuture<?> wakeUp;
+
+    // state
+    private long runAt;
+    private boolean enabled;
+
+    Rescheduler(Runnable r, ChannelExecutor exec, ScheduledExecutorService scheduler) {
```

Comment:
There is a way to decouple Rescheduler from ChannelExecutor.  Pass an `Executor` here and make it an requirement that it must serialize its runnables and `rechedule()`. In `ManagedChannelImpl`, you could make an `Executor` that delegates to `ChannelExecutor`. I am strong for splitting this class out of `ManagedChannelImpl`, which is good for test-ability and also for the fitness of `ManagedChannelImpl`.

Reply:
Done.